add(platform): BYOC networking and security (#298)
aiven/aiven-docs: docs/platform/howto/byoc/networking-security.md

| Aiven Management plane | Bastion subnet | Aiven uses SSH from the Aiven Management plane to the bastion service node for troubleshooting, setup, or configuration tasks on the bastion. |
| Aiven Management plane | Bastion subnet | Proxy service is used to tunnel the traffic from the Aiven Management plane to workload nodes.                                                |
| Aiven Management plane | Bastion subnet | Aiven Management plane accesses the bastion service node to retrieve status information and perform routine operations.                       |

### Traffic from the bastion

| Source         | Destination      | Description                                                                                                                             |
| -------------- | -----------------| --------------------------------------------------------------------------------------------------------------------------------------- |
| Bastion subnet | Aiven Management | Communication channel with the Aiven Management plane used to collect metrics and logs from the bastion and workload nodes              |

> | Bastion subnet | Aiven Management | Required for the bastion and workload nodes to make calls to the Aiven Management plane                                                 |

Author: @kevinmfox @jeremy-heller-aiven Last time we decided to keep all on separate rows. Please see @sigmaris comment. Wdyt?

> | Bastion subnet | AWS CloudFront   | Communication channel for downloading RPM packages for bastion and workload nodes setup                                                 |

Reviewer: In the standard BYOC model, RPM downloads for workload nodes don't pass through the bastion, also I think "communication channel" suggests two-way communication or telemetry, but actually it's purely one-way downloading of RPMs that happens from Cloudfront, so:

```suggestion
| Bastion subnet | AWS CloudFront   | Used for downloading RPM packages for bastion nodes setup                                                 |
```

| Bastion subnet | Workload subnet  | Aiven uses SSH from the bastion subnet to the workload subnet for troubleshooting, setup, or configuration tasks on the workload nodes. |

> | Bastion subnet | Workload subnet  | Aiven Management plane accesses the workload nodes via the bastion node to retrieve status information and perform routine operations.  |

Author: @kevinmfox @jeremy-heller-aiven again, wdyt?

| Bastion subnet | DNS & NTP        | Destination dependant on cloud provider                                                                                                 |

:::note
Aiven Management IPs are considered relatively stable but do occasionally change and are
considered dynamic for the purposes of firewall policies. To that end, an egress of

> `0.0.0.0/0` is generally required to accommodate traffic to Aiven Management.

Author: @sigmaris could you propose relevant wording for this information please?

> Aiven Management IPs are considered relatively stable but do occasionally change and are
> considered dynamic for the purposes of firewall policies. To that end, an egress of
> `0.0.0.0/0` is generally required to accommodate traffic to Aiven Management.

Reviewer: I think we could just classify them together; actually as I understand it, Aiven Management static IPs are currently somewhat more stable than Cloudfront IP ranges, but we would still need a `0.0.0.0/0` egress rule to allow for changes to Cloudfront, and the Aiven Management IPs will hopefully change to a smaller set of Aiven-owned IP space in the near future.

```suggestion
Aiven Management and Cloudfront IP ranges do occasionally change and are considered dynamic for the purposes of firewall policies. To that end, an egress of `0.0.0.0/0` is generally required to accommodate traffic to these destinations.
```

Aiven support staff require direct access to the bastion or workload nodes. SSH
connections from support staff are logged, monitored, and require validation by Aiven’s
security operations team. SSH connections originate from Aiven Management plane's NAT gateways.

### Traffic to the workload

On top of SSH and Aiven Management Plane connectivity, there are also ingress and egress
rules from the Customer Networks that need to be defined to allow customer applications to
connect to Aiven services. This depends on the cloud provider being used, and the
customer’s network architecture.

| Source                   | Destination     | Description                                                                                                                             |
| ------------------------ | ----------------| --------------------------------------------------------------------------------------------------------------------------------------- |
| Bastion subnet           | Workload subnet | Aiven uses SSH from the bastion subnet to the workload subnet for troubleshooting, setup, or configuration tasks on the workload nodes. |

> | Bastion subnet           | Workload subnet | Aiven Management plane accesses the workload nodes via the bastion node to retrieve status information and perform routine operations.  |

Reviewer: These could be grouped together and just refer back to the description of this traffic in the "Traffic from the bastion" section, I think. The same thing is described in two places.

Author: I assume not everyone is going to read throughout the whole document and some people might be only interested in a single section. We'd like to avoid making them jump back and forth to look somewhere else for info that's why including it also here for ease of accessibility.
About the grouping, again I defer to @jeremy-heller-aiven & @kevinmfox ...

| Your remote applications | Workload subnet | Varies depending on services being used.                                                                                                |

### Traffic from the workload

The workload subnet nodes communicate with each other. The nature of these communications
varies depending on the services, features, and plugins being used.

| Source          | Destination              | Description                                                                                                                |
| --------------- | -------------------------| -------------------------------------------------------------------------------------------------------------------------- |
| Workload subnet | Aiven Management         | Communication channel with the Aiven Management plane used to collect metrics and logs from the bastion and workload nodes |

> | Workload subnet | Aiven Management         | Required for the bastion and workload nodes to make calls to the Aiven Management plane                                    |

Reviewer: See above comment on the "Traffic from the bastion" table.

> | Workload subnet | AWS CloudFront           | Communication channel for downloading RPM packages for bastion and workload nodes setup                                    |

Reviewer: Traffic from the workload subnet to CloudFront is only relevant to the workload nodes:
```suggestion
| Workload subnet | AWS CloudFront           | Used for downloading RPM packages for workload nodes setup                                    |
```

| Workload subnet | Workload subnet          | Intra-subnet communication between Aiven nodes. Ports are dynamic and dependent on services, features, and plugins used.   |
| Workload subnet | Your remote applications | Varies depending on services being used.                                                                                   |
| Workload subnet | DNS & NTP                | Destination dependant on cloud provider                                                                                    |

:::note
Aiven Management IPs are considered relatively stable but do occasionally change and are
considered dynamic for the purposes of firewall policies. To that end, an egress of

> `0.0.0.0/0` is generally required to accommodate traffic to Aiven Management.

Reviewer: FWIW this equally applies to AWS Cloudfront IPs, we could also want to mention that as well here.

:::

## Security and compliance

### System security

The bastion and workload nodes leverage hardened images customized to only include
components required for the operation of Aiven services.

All nodes use full-disk encryption specific to the customer and service, and are rotated
during maintenance events. Encryption keys are stored securely in an Aiven proprietary KMS,
and only specific and approved operators have access to this system. All access is logged
and monitored.

If required for troubleshooting or incident investigation, SSH connections to the nodes
can only be performed by specific and approved operators, and can only originate from

> specific Aiven NAT gateway IPs. All access is logged and monitored, and Aiven’s security team

Reviewer: ```suggestion
specific Aiven gateway IPs. All access is logged and monitored, and Aiven’s security team
```

---
title: Bring your own cloud networking and security
sidebar_label: Networking & security
---

Aiven combines a multi-cloud strategy with a cloud-agnostic approach to make the [bring your own cloud (BYOC)](/docs/platform/concepts/byoc) experience not only versatile and cost-efficient but also secure.

## Bastion proxy networking

The [bastion](https://en.wikipedia.org/wiki/Bastion_host) proxy service acts as a trusted
relay between the Aiven Management plane and your BYOC environment. A typical BYOC
implementation results in workloads that are not directly accessible from external networks.

> The bastion proxy is located in a separate virtual private cloud (VPC) and facilitates the

Reviewer 2: Good point, on AWS and GCP we now have a 1 VPC model with 2 subnets. I believe Azure can work the same way, but needs to be confirmed. @kevinmfox we should make sure both documents specify the 1 VPC model

communication between Aiven Management and monitoring systems and your private workloads.
The bastion proxy service is also responsible for facilitating a communication channel that
allows the Aiven Management plane to launch services within your private cloud.

### Traffic to the bastion

> Aiven leverages redundant gateways to conduct BYOC operations. Ingress traffic from
> particular Aiven gateway IP addresses to the bastion subnet on specific ports is required.
> Each gateway is hosted on a separate and geographically remote cloud provider to protect
> against communication failures.

Reviewer 3: ```suggestion
Aiven Management plane traffic to your BYOC environment originates from redundant NAT gateways. The bastion subnet therefore needs to whitelist 4 static IPs on the ports listed below. 
```

| Source                 | Destination    | Description                                                                                                                                   |
| ---------------------- | ---------------| --------------------------------------------------------------------------------------------------------------------------------------------- |
| Aiven Management plane | Bastion subnet | Aiven uses SSH from the Aiven Management plane to the bastion service node for troubleshooting, setup, or configuration tasks on the bastion. |
| Aiven Management plane | Bastion subnet | Proxy service is used to tunnel the traffic from the Aiven Management plane to workload nodes.                                                |

> | Aiven Management plane | Bastion subnet | Aiven Management plane accesses the bastion service node to retrieve status information and perform routine operations.                       |

Reviewer 3: Shouldn't this table include the ports we require?

Reviewer 2: Kevin Fox has a deep dive doc that lists all the ports and IPs, we can exclude the ports in the public version

### Traffic from the bastion

| Source         | Destination      | Description                                                                                                                             |
| -------------- | -----------------| --------------------------------------------------------------------------------------------------------------------------------------- |
| Bastion subnet | Aiven Management | Communication channel with the Aiven Management plane used to collect metrics and logs from the bastion and workload nodes              |
| Bastion subnet | Aiven Management | Required for the bastion and workload nodes to make calls to the Aiven Management plane                                                 |

> | Bastion subnet | AWS CloudFront   | Communication channel for downloading RPM packages for bastion and workload functionality and maintenance                               |

Reviewer 3: ```suggestion
| Bastion subnet | AWS CloudFront   | Communication channel for downloading RPM packages for bastion node setup                               |
```

| Bastion subnet | Workload subnet  | Aiven uses SSH from the bastion subnet to the workload subnet for troubleshooting, setup, or configuration tasks on the workload nodes. |
| Bastion subnet | Workload subnet  | Aiven Management plane accesses the workload nodes via the bastion node to retrieve status information and perform routine operations.  |

> | Bastion subnet | DNS & NTP        | Destination dependant on cloud provider                                                                                                 |

Reviewer 3: Same here, could even specify source and destination ports where it makes sense

Reviewer 4: Without port numbers, some descriptions look a bit redundant. Eg. compare lines 38 and 39.

Author: See https://github.com/aiven/aiven-docs/pull/298/files#r1657455784

:::note
Aiven Management and CloudFront IP ranges do occasionally change and are considered dynamic
for the purposes of firewall policies. To that end, an egress of `0.0.0.0/0` is generally

> required to accommodate traffic to these destinations.

Reviewer 5: Just a suggestion, maybe we can simplify this a bit: 

Aiven Management and CloudFront IP ranges occasionally change and are considered dynamic for firewall policies. To accommodate traffic to these destinations, an egress of 0.0.0.0/0 is generally required.

`0.0.0.0/0` is generally required to accommodate traffic to Aiven Management.
:::

## Workload nodes networking

> Management traffic to and from backend workload nodes is sent through the bastion proxy and

Reviewer 3: ```suggestion
Management traffic to and from workload nodes is sent through the bastion proxy and
```

encrypted. Aiven Management and API calls are always encrypted. SSH access is permitted if

> Aiven support staff require direct access to the bastion service or workload nodes. SSH

Reviewer 3: ```suggestion
Aiven support staff require direct access to the bastion or workload nodes. SSH
```

connections from support staff are logged, monitored, and require validation by Aiven’s

> security operations team. SSH connections originate from the redundant Aiven gateways.

Reviewer 3: ```suggestion
security operations team. SSH connections originate from Aiven Management plane's NAT gateways.
```

Reviewer 4: Do we really need to say that Aiven's gateways are NAT gateways? I even think they are some in-house `crabredir` proxies.

Reviewer: Aiven's gateways (`gw1` and `gw2`) are not NAT gateways by the commonly understood meaning of that term. You could argue that they tunnel connections through proxy programs to 'translate' the source address of the connections somehow, but that'd just confuse the majority of the audience who would understand NAT as packet-level IPv4 Network Address Translation.

| Your remote applications | Workload subnet | Varies depending on services being used.                                                                                                |

### Traffic from the workload

The workload subnet nodes communicate with each other. The nature of these communications
varies depending on the services, features, and plugins being used.

| Source          | Destination              | Description                                                                                                                |
| --------------- | -------------------------| -------------------------------------------------------------------------------------------------------------------------- |
| Workload subnet | Aiven Management         | Communication channel with the Aiven Management plane used to collect metrics and logs from the bastion and workload nodes |
| Workload subnet | Aiven Management         | Required for the bastion and workload nodes to make calls to the Aiven Management plane                                    |

> | Workload subnet | AWS CloudFront           | Communication channel for downloading RPM packages for bastion and workload functionality and maintenance                  |

Reviewer 3: ```suggestion
| Workload subnet | AWS CloudFront           | Communication channel for downloading RPM packages for workload node setup                  |
```

| Workload subnet | DNS & NTP                | Destination dependant on cloud provider                                                                                    |

:::note
Aiven Management and CloudFront IP ranges do occasionally change and are considered dynamic
for the purposes of firewall policies. To that end, an egress of `0.0.0.0/0` is generally
required to accommodate traffic to these destinations.
:::

## Security and compliance

### System security

The bastion and workload nodes leverage hardened images customized to only include
components required for the operation of Aiven services.

All nodes use full-disk encryption specific to the customer and service, and are rotated
during maintenance events. Encryption keys are stored securely in an Aiven proprietary KMS,
and only specific and approved operators have access to this system. All access is logged
and monitored.

If required for troubleshooting or incident investigation, SSH connections to the nodes
can only be performed by specific and approved operators, and can only originate from
specific Aiven gateway IPs. All access is logged and monitored, and Aiven’s security team

> follows-up on access to ensure approval and validity.

Reviewer 5: nit: should follows-up be hyphenated when used as a verb?

> specific Aiven gateway IPs. All access is logged and monitored, and Aiven’s security team

Reviewer 3: ```suggestion
specific Aiven NAT gateway IPs. All access is logged and monitored, and Aiven’s security team
```
